src/MongoDB.Driver/Linq/MongoQueryable.cs

```diff
@@ -787,72 +773,15 @@ public static Task<TSource> FirstOrDefaultAsync<TSource>(this IMongoQueryable<TS
-        {
-            Ensure.IsNotNull(source, nameof(source));
-            Ensure.IsNotNull(keySelector, nameof(keySelector));
-            Ensure.IsNotNull(resultSelector, nameof(resultSelector));
-
-            return (IMongoQueryable<TResult>)Queryable.GroupBy(source, keySelector, resultSelector);
-        }
-
-        /// <summary>
-        /// Correlates the elements of two sequences based on key equality and groups the results.
+        /// 
         /// </summary>
-        /// <typeparam name="TOuter">The type of the elements of the first sequence.</typeparam>
-        /// <typeparam name="TInner">The type of the elements of the second sequence.</typeparam>
-        /// <typeparam name="TKey">The type of the keys returned by the key selector functions.</typeparam>
-        /// <typeparam name="TResult">The type of the result elements.</typeparam>
-        /// <param name="outer">The first sequence to join.</param>
-        /// <param name="inner">The sequence to join to the first sequence.</param>
-        /// <param name="outerKeySelector">A function to extract the join key from each element of the first sequence.</param>
-        /// <param name="innerKeySelector">A function to extract the join key from each element of the second sequence.</param>
-        /// <param name="resultSelector">A function to create a result element from an element from the first sequence and a collection of matching elements from the second sequence.</param>
-        /// <returns>
-        /// An <see cref="IMongoQueryable{TResult}" /> that contains elements of type <typeparamref name="TResult" /> obtained by performing a grouped join on two sequences.
-        /// </returns>
-        public static IMongoQueryable<TResult> GroupJoin<TOuter, TInner, TKey, TResult>(this IMongoQueryable<TOuter> outer, IEnumerable<TInner> inner, Expression<Func<TOuter, TKey>> outerKeySelector, Expression<Func<TInner, TKey>> innerKeySelector, Expression<Func<TOuter, IEnumerable<TInner>, TResult>> resultSelector)
+        /// <typeparam name="TDocument"></typeparam>
+        /// <param name="source"></param>
+        /// <returns></returns>
+        public static BsonDocument[] GetLoggedStages<TDocument>(this IQueryable<TDocument> source)
         {
-            Ensure.IsNotNull(outer, nameof(outer));
-            Ensure.IsNotNull(inner, nameof(inner));
-            Ensure.IsNotNull(outerKeySelector, nameof(outerKeySelector));
-            Ensure.IsNotNull(innerKeySelector, nameof(innerKeySelector));
-            Ensure.IsNotNull(resultSelector, nameof(resultSelector));
-
-            return (IMongoQueryable<TResult>)Queryable.GroupJoin(outer, inner, outerKeySelector, innerKeySelector, resultSelector);
+            var provider = GetMongoQueryProvider(source, nameof(GetLoggedStages));
```

Comment:
If `source.Provider` (which is of type `IQueryProvider`) does not also implement `IMongoQueryProvider` this method will throw an exception.